modules/basic/ds/hashmap.h

```diff
@@ -217,6 +225,262 @@ class HashmapBuilder : public HashmapBaseBuilder<K, V, H, E> {
+    return nullptr;
+  }
+
+  /**
+   * @brief Return the const ending iterator.
+   *
+   */
+  void* cend() const {
+    // return hashmap_.cend();
+    LOG(INFO) << __func__ << " is not finished yet.";
+    return nullptr;
+  }
+
+  /**
+   * @brief Find the value by key.
+   *
+   */
+  void* find(const K& key) {
+    // return hashmap_.find(key);
+    LOG(INFO) << __func__ << " is not finished yet.";
+    return nullptr;
+  }
+
+  /**
+   * @brief Associated with a given data buffer
+   */
+  void AssociateDataBuffer(std::shared_ptr<Blob> data_buffer) {
+    // this->data_buffer_ = data_buffer;
+  }
+
+  template <typename K_ = K>
+  typename std::enable_if<std::is_integral<K_>::value, void>::type Construct(
+      int concurrency = 1) {
+    size_t count = 0;
+    vec_kv_.resize(n_elements_);
+    vec_k_.resize(vec_kv_.size());
+    uint64_t start_time = GetCurrentTime();
+    for (auto& kv_ : vec_kv_) {
+      vec_k_[count] = kv_.first;
+      count++;
+    }
+    LOG(INFO) << "Constructing the vec_k_ takes "
+              << GetCurrentTime() - start_time << " s.";
+
+    auto data_iterator = boomphf::range(vec_k_.begin(), vec_k_.end());
+    auto bphf = boomphf::mphf<K, hasher_t>(vec_k_.size(), data_iterator,
+                                           concurrency, 1.0f);
+
+    vec_v_.resize(count);
+    count = vec_k_.size() / concurrency;
+    start_time = GetCurrentTime();
+    parallel_for(
+        0, concurrency,
+        [&](const int i) {
+          if (unlikely(i == concurrency - 1)) {
+            for (size_t j = i * count; j < vec_v_.size(); j++) {
+              vec_v_[bphf.lookup(vec_k_[j])] = vec_kv_[j].second;
+            }
+          } else {
+            for (size_t j = i * count; j < (i + 1) * count; j++) {
+              vec_v_[bphf.lookup(vec_k_[j])] = vec_kv_[j].second;
+            }
+          }
+        },
+        concurrency);
+    LOG(INFO) << "Parallel for constructing the vec_v_ takes "
+              << GetCurrentTime() - start_time << " s.";
+  }
+
+  template <typename K_ = K>
+  typename std::enable_if<!std::is_integral<K_>::value, void>::type Construct(
+      int concurrency = 1) {
+    LOG(INFO) << __func__ << " is not supported yet.";
+  }
+
+  /**
+   * @brief Build the hashmap object.
+   *
+   */
+  Status Build(Client& client) override {
+    Construct(currency_);
+
+    auto ph_values_builder =
+        std::make_shared<ArrayBuilder<V>>(client, vec_v_.data(), vec_v_.size());
+
+    this->set_num_elements_(vec_kv_.size());
+
+    this->set_ph_values_(
+        std::static_pointer_cast<ObjectBase>(ph_values_builder));
+
+    return Status::OK();
+  }
+
+ private:
+  std::vector<V> vec_v_;
+  std::vector<K> vec_k_;
+  std::vector<std::pair<K, V>> vec_kv_;
+  uint64_t n_elements_ = 0;
+#if defined(__linux__) || defined(__linux) || defined(linux) || \
+    defined(__gnu_linux__)
+  int currency_ = sysconf(_SC_NPROCESSORS_ONLN);
+#else
+  int currency_ = 16;
```

Comment:
What is `currency_` ? If you mean `concurrency_`, defaults it to `std::thread::hardware_concurrency()` please.

Reply:
Done